src/components/Pagination/index.tsx

```diff
@@ -0,0 +1,474 @@
+    goToFirstPage,
+    goToLastPage,
+    goToNextPage,
+    goToPage,
+    goToPreviousPage,
+    maxPage,
+    page: currentPage,
+    pageData,
+    paginatedData,
+    isLoadingPage,
+    perPage,
+    reloadPage,
+    setPageData,
+    setPaginatedData,
+  } = usePagination({
+    data,
+    onChangePage: handleChangePage,
+    onLoadPage: onLoadPage ? handleLoadPage : undefined,
+    page,
+    pageCount,
+    perPage: perPageProp,
+  })
+
+  useEffect(() => {
+    if (pageProp && pageProp !== page) {
+      setPage(pageProp)
+    }
+  }, [pageProp, page])
+
+  useEffect(() => {
+    if (dataProp) {
+      setData(dataProp)
+    }
+  }, [dataProp])
+
+  const value = useMemo(() => {
+    paginationRef.current = {
+      canLoadMore: onLoadPageRef.current !== undefined,
+      data: dataProp,
+      goToFirstPage,
+      goToLastPage,
+      goToNextPage,
+      goToPage,
+      goToPreviousPage,
+      isLoadingPage,
+      maxPage,
+      page: currentPage,
+      pageData,
+      paginatedData,
+      perPage,
+      reloadPage,
+      setPageData,
+      setPaginatedData,
+    }
+
+    return paginationRef.current
+  }, [
+    goToFirstPage,
+    goToLastPage,
+    goToNextPage,
+    goToPage,
+    goToPreviousPage,
+    isLoadingPage,
+    dataProp,
+    maxPage,
+    currentPage,
+    pageData,
+    paginatedData,
+    perPage,
+    reloadPage,
+    setPageData,
+    setPaginatedData,
+  ])
+
+  useEffect(() => {
+    onLoadPageRef.current = onLoadPage
+  }, [onLoadPage])
+
+  useEffect(() => {
+    onChangePageRef.current = onChangePage
+  }, [onChangePage])
+
+  const TypedPaginationContainer = PaginationContainer as FunctionComponent<
+    PaginationContainerProps<T>
+  >
+
+  return (
+    <PaginationContext.Provider value={value as PaginationContextValue}>
+      {isLoadingPage ? (
+        <LoaderComponent paginationState={value} pageTabCount={pageTabCount} />
+      ) : null}
+      {children && typeof children === 'function' && !isLoadingPage
+        ? children(value)
+        : null}
+      {children && typeof children !== 'function' && !isLoadingPage
+        ? children
+        : null}
+      <TypedPaginationContainer
+        paginationState={value}
+        pageTabCount={pageTabCount}
+        LeftComponent={LeftComponent}
+        RightComponent={RightComponent}
+        MiddleComponent={MiddleComponent}
+      />
+    </PaginationContext.Provider>
+  )
+}
+
+type PaginationType<T = unknown> = ForwardRefExoticComponent<
+  PaginationProps<T> & RefAttributes<PaginationState<T>>
+> & {
+  RightComponent: FunctionComponent<PaginationComponentProps<T>>
+  LeftComponent: FunctionComponent<PaginationComponentProps<T>>
+  PaginationContainer: FunctionComponent<PaginationComponentProps<T>>
+  MiddleComponent: FunctionComponent<PaginationComponentProps<T>>
+  LoaderComponent: FunctionComponent<PaginationComponentProps<T>>
+}
+
+const Pagination = forwardRef(PaginationForwardFn) as PaginationType
+
+Pagination.propTypes = {
```

Comment:
Put the propTypes on the component, not the forwardRef function :)

Reply:
According to https://reactjs.org/docs/typechecking-with-proptypes.html : proptypes must be defined on the forwardRef return (ForwardRefExoticComponent interface include propTypes and defaultProps). To make sure it was right, i tried to put propTypes on PaginationForwardFn and the ts compiler throw me an error. I think i will keep propTypes on forwardRef return. 😅